src/dsa/rpo_falcon512/keys.rs

```diff
@@ -113,87 +101,128 @@ impl KeyPair {
     pub fn public_key(&self) -> PublicKey {
         // TODO: memoize public key commitment as computing it requires quite a bit of hashing.
         // expect() is fine here because we assume that the key pair was constructed correctly.
-        PublicKey::new(self.public_key).expect("invalid key pair")
-    }
-
-    /// Returns the expanded public key corresponding to this key pair.
-    pub fn expanded_public_key(&self) -> PublicKeyBytes {
-        self.public_key
+        PublicKey::new(self.pub_key())
     }
 
     // SIGNATURE GENERATION
     // --------------------------------------------------------------------------------------------
 
-    /// Signs a message with a secret key and a seed.
+    /// Signs a message with the secret key.
     ///
     /// # Errors
     /// Returns an error of signature generation fails.
-    #[cfg(feature = "std")]
     pub fn sign(&self, message: Word) -> Result<Signature, FalconError> {
-        let msg = message.iter().flat_map(|e| e.as_int().to_le_bytes()).collect::<Vec<_>>();
-        let msg_len = msg.len();
-        let mut sig = [0_u8; SIG_LEN + NONCE_LEN];
-        let mut sig_len: usize = 0;
-
-        let res = unsafe {
-            ffi::PQCLEAN_FALCON512_CLEAN_crypto_sign_signature_rpo(
-                sig.as_mut_ptr(),
-                &mut sig_len as *mut usize,
-                msg.as_ptr(),
-                msg_len,
-                self.secret_key.as_ptr(),
-            )
+        let mut rng = thread_rng();
+        let mut nonce = [0u8; SIG_NONCE_LEN];
+        rng.fill_bytes(&mut nonce);
```

Comment:
One of the main reasons for implementing signing in pure Rust is so that we can generate signatures in WASM. This won't work if we need to use `thread_rng()`. There are a few ways to address this - the simplest one is probably to split this into two methods:
```Rust
pub fn sign(&self, message: Word) -> Result<Signature, FalconError> {
    let mut rng = thread_rng();
    let mut nonce = [0u8; SIG_NONCE_LEN];
    rng.fill_bytes(&mut nonce);
    self.sign_with_nonce(message, nonce)
}

pub fn sign_with_nonce(&self, message: Word, nonce: Nonce) -> Result<Signature, FalconError> {
    todo!()
}
```
Then, `sign()` would be available only in `std` context, but `sign_with_nonce()` would be available always.

There are other ways to handle this as well (e.g., via traits) - but let's start out with the simple one.

Reply:
The issue is that the signing algorithm uses randomness in an essential way through the FF sampling. Thus without a randomness generator I am not sure we will be able to generate signatures even if we are given the nonce.

Reply:
Ah - I missed this. I guess we'll need to go to the trait method sooner rather than later. There are a couple of questions here:

1. Should we parametrize just the `sign` method or the whole struct? (maybe start out with the `sign()` method for now).
2. Should we use `RngCore` trait - or come up with a simple trait of our own (e.g., something similar to `FeltRng`)?

Reply:
1. We can start with `sign` and then propagate.
2. I think it is sufficient to use `RngCore`.

Would the following be acceptable?
```rust
pub fn sign(&self, message: Word, rng: &mut dyn RngCore) -> Result<Signature, FalconError> {...}

let mut rng = OsRng;
let signature = sk.sign(message, &mut rng);
```

Reply:
I think this could work. The main question is how difficult would it be to provide a custom implementation of `RngCore` (e.g., could we implement it against `RpoRandomCoin`)?

Reply:
But then we would have to worry about it being cryptographically secure, right?

Reply:
You mean `RpoRandomCoin`? Wouldn't we need to worry about this either way? :) (since we use it to generate randomness in the STARK protocol).

Reply:
I agree :)
Actually I should've asked why would we need to implement it over `RpoRandomCoin`? At the end, the goal is to get bytes from a good source of entropy.

Reply:
`RpoRandomCoin` was just an example here. But we may want to provide an implementation which generates randomness deterministically from some seed. Or maybe works by calling some external function in WASM context to get random values. So, the main question is how difficult it is to implement `RngCore` trait (i.e., if it is difficult, we could create our own trait with reduced functionality required only for our use case).

Reply:
Got it. 
This is how difficult it is to implement `RngCore`:

```rust
use rand_core::{impls::{next_u32_via_fill, next_u64_via_fill}, OsRng, RngCore};

    pub struct RngMiden {}

    impl RngCore for RngMiden {
        fn next_u32(&mut self) -> u32 {
            next_u32_via_fill(self)
        }
    
        fn next_u64(&mut self) -> u64 {
            next_u64_via_fill(self)
        }
    
        fn fill_bytes(&mut self, dest: &mut [u8]) {
            todo!()
        }
    
        fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand::Error> {
            Ok(self.fill_bytes(dest))
        }
    }
```

So the core of the implementation is in `fill_bytes`.